app/models/ability.rb

```diff
@@ -4,22 +4,34 @@ class Ability
   include CanCan::Ability
 
   def initialize(user)
-    # Anyone can view projects not owner by a user or a school.
+    define_common_non_student_abilities(user)
+
+    return unless user
+
+    define_authenticated_non_student_abilities(user)
+
+    user.schools.each do |school|
+      define_school_student_abilities(user:, school:) if user.school_student?(school)
+      define_school_teacher_abilities(user:, school:) if user.school_teacher?(school)
+      define_school_owner_abilities(school:) if user.school_owner?(school)
+    end
+  end
+
+  private
+
+  def define_common_non_student_abilities(user)
+    return if user&.student?
+
+    # Anyone can view projects not owned by a user or a school.
     can :show, Project, user_id: nil, school_id: nil
     can :show, Component, project: { user_id: nil, school_id: nil }
 
     # Anyone can read publicly shared lessons.
     can :read, Lesson, visibility: 'public'
+  end
 
-    return unless user
-
-    # Any authenticated user can create projects not owned by a school.
-    can :create, Project, user_id: user.id, school_id: nil
-    can :create, Component, project: { user_id: user.id, school_id: nil }
-
-    # Any authenticated user can manage their own projects.
-    can %i[read update destroy], Project, user_id: user.id
-    can %i[read update destroy], Component, project: { user_id: user.id }
+  def define_authenticated_non_student_abilities(user)
+    return if user&.student?
 
     # Any authenticated user can create a school. They agree to become the school-owner.
     can :create, School
```

Comment:
Shouldn't we ensure any abilities beyond this point are only given to a teacher or owner? else any one with an account could theoretically do these things - unlikely as that may be

Reply:
Ah ok so we do want anyone to be able to do these things. We explicitly don't want students to be create a school (i.e. the school onboarding flow) or create projects/components that aren't attached to a school. Everyone else, teacher, owners, non-school members can do all those things